src/main/scala/firrtl/transforms/Dedup.scala

```diff
@@ -18,134 +19,301 @@ case class NoDedupAnnotation(target: ModuleName) extends SingleTargetAnnotation[
 
-    val cname = CircuitName(c.main)
-    val renameMap = RenameMap(dedupMap.map { case (from, to) =>
-      logger.debug(s"[Dedup] $from -> $to")
-      ModuleName(from, cname) -> List(ModuleName(to, cname))
-    })
+    // Maps a module's tag to its deduplicated module
+    val tag2name = mutable.HashMap.empty[String, String]
+
+    // Maps a module's name to its tag
+    val name2tag = mutable.HashMap.empty[String, String]
+
+    // Maps a tag to all matching module names
+    val tag2all =  mutable.HashMap.empty[String, mutable.Set[String]]
+
+    // Build dedupMap
+    moduleLinearization.foreach { moduleName =>
+      // Get original module
+      val originalModule = moduleMap(moduleName)
+
+      // Replace instance references to new deduped modules
+      val dontcare = RenameMap()
+      dontcare.setCircuit("dontcare")
+      //val fixedModule = DedupModules.dedupInstances(originalModule, tag2module, name2tag, name2module, dontcare)
+
+      if(noDedups.contains(originalModule.name)) {
+        // Don't dedup. Set dedup module to be the same as fixed module
+        name2tag(originalModule.name) = originalModule.name
+        tag2name(originalModule.name) = originalModule.name
+        //templateModules += originalModule.name
+      } else { // Try to dedup
+
+        // Build name-agnostic module
+        val agnosticModule = DedupModules.agnostify(originalModule, name2tag, tag2name)
+
+        // Build tag
+        val tag = md5Hash(agnosticModule match {
+          case Module(i, n, ps, b) =>
+            ps.map(_.serialize).mkString + b.serialize
+          case ExtModule(i, n, ps, dn, p) =>
+            ps.map(_.serialize).mkString + dn + p.map(_.serialize).mkString
+        })
+
+        // Match old module name to its tag
+        name2tag(originalModule.name) = tag
+
+        // Set tag's module to be the first matching module
+        if(!tag2name.contains(tag)) {
+          tag2name(tag) = originalModule.name
+          tag2all(tag) = mutable.Set(originalModule.name)
+        } else {
+          tag2all(tag) += originalModule.name
+        }
+      }
+    }
 
-    (c.copy(modules = dedupedModules), renameMap)
+
+    // Set tag2name to be the best dedup module name
+    val moduleIndex = circuit.modules.zipWithIndex.map{case (m, i) => m.name -> i}.toMap
+    def order(l: String, r: String): String = if(moduleIndex(l) < moduleIndex(r)) l else r
+    tag2all.foreach { case (tag, all) => tag2name(tag) = all.reduce(order)}
+
+    // Create map from original to dedup name
+    val name2name = name2tag.map{case (name, tag) => name -> tag2name(tag)}
+
+    // Build Remap for modules with deduped module references
+    val tag2module = tag2name.map { case (tag, name) => tag -> DedupModules.dedupInstances(name, moduleMap, name2name, renameMap) }
+
+    // Build map from original name to corresponding deduped module
+    val name2module = name2tag.map {case (name, tag) => name -> tag2module(tag)}
```

Comment:
Would appreciate consistent spacing around the curly braces, all 3 of these use different spacing. I generally prefer `collection.map { case (...) => }` although @albert-magyar might be slowly converting me to `collection.map({ case (...) => })`, jury is still out on that one.